### DISCUSSION TITLE
Pydantic usage

### DISCUSSION THREAD

**Discussion by akaviaLab** [opencobra/cobrapy#1196] [Ideas]
I've been going over the cobra issues, and trying to find issues/pull request I can tackle.
I saw #959 where @Midnighter mentioned that he would prefer Pydantic models. I'm not sure how exactly you'd like to use them.

It seems that
1) Rewrite dict and json to use pydantic. So Metaoblite would be something like
```
from typing import Optional
from pydantic import BaseModel

class MetaboliteDict(BaseModel):
    id: int
    name: str
    compartment: str
    charge: Optional[float] = None
    formula: Optional[str] = None
    _bound: Optional[float] = 0
    notes: Optional[dict] = {}
    annotation: Optional[dict]= {}
```

and then rely on pydantic to validate and convert from cobra.Metaoblite. reactionDict will be more complicated, since there are more validations and conversion.
json.py will rely on dict.py and use Pydantic to convert to/from JSON

2) Rewrite the base classes to include Pydantic. I'm not sure how that would work or if that makes sense.

Is (1) correct? Is the format something like what you were thinking?

**Comment by cdiener** [comment]
There is more discussion in https://github.com/opencobra/cobrapy/issues/967 . Though we probably want to hold off with that for after the 1.0 release and the paper. Though if you are looking for some issues to tackle I would love some help with #647 or #1040.

**Reply by Midnighter** [reply]
I agree with @cdiener here. This is best tackled in a move to implement #967. If that tickles your fancy, the first step will be to agree on a definition of a constraint-based model as a subset of the SBML models.

**Reply by matthiaskoenig** [reply]
This should be done consistently on the complete code base. I would not start this now as part of another issue. This will have major implications for many things.

**Reply by akaviaLab** [reply]
I'm going to wait on this until it is more defined. I'll probably tackle #1040 or something like that.
